src/Runner.Worker/Container/ContainerHookManager.cs

```diff
@@ -0,0 +1,158 @@
+        {
+            Trace.Entering();
+
+            var meta = new ContainerHookMeta
+            {
+                Command = "cleanup_job", // GetHookCommand(nameof(CleanupJobAsync)),
+                ResponseFile = "response.json",
+                Args = new ContainerHookArgs
+                {
+                    JobContainer = containers.Where(c => c.IsJobContainer).FirstOrDefault().GetHookContainer(),
+                    Services = containers.Where(c => c.IsJobContainer == false).Select(c => c.GetHookContainer()).ToList(),
+                    Network = containers.Where(c => !string.IsNullOrEmpty(c.ContainerNetwork)).FirstOrDefault()?.ContainerNetwork,
+                }
+            };
+            return await ExecuteHookScript(context, GetHookIndexPath(), meta);
+        }
+
+        public async Task<int> ContainerStepAsync(IExecutionContext context)
+        {
+            Trace.Entering();
+            await Task.FromResult(0);
+            throw new NotImplementedException();
+        }
+
+        public async Task<int> RunScriptStepAsync(IExecutionContext context)
+        {
+            Trace.Entering();
+            await Task.FromResult(0);
+            throw new NotImplementedException();
+        }
+
+        private async Task<int> ExecuteHookScript(IExecutionContext context, string hookScriptPath, ContainerHookMeta args)
+        {
+            var scriptDirectory = Path.GetDirectoryName(hookScriptPath);
+            var stepHost = HostContext.CreateService<IDefaultStepHost>();
+            Dictionary<string, string> inputs = new()
+            {
+                ["standardInInput"] = JsonUtility.ToString(args),
+                ["path"] = hookScriptPath,
+                ["shell"] = Path.Combine(HostContext.GetDirectory(WellKnownDirectory.Externals), NodeUtil.GetInternalNodeVersion(), "bin", $"node{IOUtil.ExeExtension}") + " {0}" // TODO: fix hardcoded node path
+            };
+
+            var handlerFactory = HostContext.GetService<IHandlerFactory>();
+            var handler = handlerFactory.Create(
+                            context,
+                            null,
+                            stepHost,
+                            new ScriptActionExecutionData(),
+                            inputs,
+                            environment: new Dictionary<string, string>(VarUtil.EnvironmentVariableKeyComparer),
+                            context.Global.Variables,
+                            actionDirectory: scriptDirectory,
+                            localActionContainerSetupSteps: null) as ScriptHandler;
+            handler.PrepareExecution(ActionRunStage.Pre); // TODO: find out stage, we only use Start in pre, but double check
+            await handler.RunAsync(ActionRunStage.Pre);
+            return handler.ExecutionContext.Result == TaskResult.Failed || handler.ExecutionContext.Result == TaskResult.Canceled ? 1 : 0;
```

Comment:
nit: does this function actually need to return an exit code, we are already setting the result on the ExecutionContext, which contains this same information.